---
title: "Arabidopsis_promoters_clean"
author: "Brestovitsky A"
date: "4 February 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(rtracklayer)
library(GenomicRanges)
library(tidyverse)
```


```{r}
Annotation <- import.gff3("Arabidopsis_thaliana.TAIR10.27.gff3")
chrom_size_Ath_TAIR10 <- read_tsv("chrom_size_Ath_TAIR10.txt", col_names = FALSE)
seqlengths(Annotation) <- as.vector(chrom_size_Ath_TAIR10[[2]]) #Issential info for following calculations
seqlengths(Annotation)
#       1        2        3        4        5       Mt       Pt 
# 30934802 20026597 23850831 18894810 27425097   373044   157057 

prom_size <- 4500 #Desired promoter size
minSize <- 500 #Minimal promoter size


# Define genes, No TE
genes <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "protein_coding",]

```



```{r}
# Define all potential promoters and trim out of border regions
promoters_4500 <- trim(GenomicRanges::promoters(genes, upstream = prom_size, downstream = 0))

# filter out all potentialpromoters which include genes, and change size of promoters if genes are too close, filter for short promoters (lee than minimal length):
# 1------find gaps bteween genes
gaps_genes <- GenomicRanges::gaps(genes)

#------find overlap between promoters and gaps between genes, the shortest will become a new promoter.
intersect_promoters_gaps <- GenomicRanges::intersect(promoters_4500, gaps_genes)

# Filter short regions in potential promoter
intersect_promoters_gaps_filter <- intersect_promoters_gaps[width(intersect_promoters_gaps) >= minSize]


#------FInd genes following promoters
#"precede(query, subject)" returns the index of the interval in subject that is directly preceded by the "query" range. Overlaping regions are excluded
#gene_list <- genes[precede(intersect_promoters_gaps, genes)]
gene_list <- genes[precede(intersect_promoters_gaps_filter, genes)]
length(gene_list) == length(intersect_promoters_gaps_filter) #should be TRUE

#------Update Metadata of intersect_promoters_gaps, and change "ID", "type" and "biotype" columns to be "promoter". "ID" should be unique identifier!!
mcols(intersect_promoters_gaps_filter) <- mcols(gene_list) 

mcols(intersect_promoters_gaps_filter)$type <- "promoter"
mcols(intersect_promoters_gaps_filter)$biotype <- "promoter"
mcols(intersect_promoters_gaps_filter)$ID <- str_replace(mcols(intersect_promoters_gaps_filter)$ID, "gene", "promoter")

 #------Sort and combine to final list for genes and promoters 
genes_promoters <- sort(c(gene_list, intersect_promoters_gaps_filter), ignore.strand=TRUE)


#Save annatation files for promoters and gene_promoters
rtracklayer::export.gff3(intersect_promoters_gaps_filter, "promoters_4500.gff3", format = "GFF3")
rtracklayer::export.gff3(genes_promoters, "genes_promoters_4500.gff3", format = "GFF3")

#End
```

```{r}
promoters
ranges(promoters)
granges(promoters)
test <- as.data.frame(granges(promoters))
paste(paste(test$seqnames, test$start, sep = ":"),test$end, sep = ("-"))


```







