---
title: "Arabidopsis_promoters"
author: "Brestovitsky A"
date: "4 February 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(rtracklayer)
library(GenomicRanges)
source("scripts/plotRanges.R")
#useful function to plot Ranges
plotRanges

```

```{r}
Annotation <- import.gff3("Arabidopsis_thaliana.TAIR10.27.gff3")
seqlengths(Annotation) <- as.vector(chrom_size_Ath_TAIR10) #Issential info for following calculations
# seqlengths(Annotation)
#       1        2        3        4        5       Mt       Pt 
# 30934802 20026597 23850831 18894810 27425097   373044   157057 

names(Annotation) <- 
#quick check up
summary(mcols(Annotation))
granges(Annotation) # left table
seqnames(Annotation) #summary of chromasomes
head(Annotation, n=10)
ranges(Annotation) #start, end, width
mcols(Annotation) #right table
names(mcols(Annotation))
table(mcols(Annotation)$biotype)
table(mcols(Annotation)$type)
table(mcols(Annotation)$source)
mcols(Annotation)$Parent[1:4]
table(mcols(Annotation)$ID)
```


```{r}
# 1. Creat small data set

#1. Define genes
genes <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "protein_coding",]
#check
table(mcols(genes)$type)
table(mcols(genes)$biotype)

#Set names to granges
names(genes) <- 1:length(start(genes))

mcols(genes)$names_GRange <- as.data.frame(names(genes))
names(mcols(genes))
mcols(genes)$gene_start <- start(genes)
mcols(genes)$gene_endend <- end(genes)


Small_genes <- genes[1:10]
GRanges(Small_genes)

```

```{r}

# Define all promoters upstream
#up_subject <- promoters(subject, upstream = 5, downstream = 0)

#----- promoter sequence (1.5kbp, 3kbp, and 4.5kbp upstream
prom_size <- 4500
minSize <- 500

#2. Define all_promoters
promoters_4500 <- trim(GenomicRanges::promoters(Small_genes, upstream = prom_size, downstream = 0))
trim(promoters_4500)

#find gaps bteween genes
gaps_genes <- GenomicRanges::gaps(Small_genes)

start = 1, end = 49286
end=seqlengths(x)
table(width(gaps_genes))
#225   465   880  2911  3630  4351  5927 13449 17246 17455 
#    1     1     1     1     1     1     1     1     1     1 
    
#find overlap between promoters and gaps between genes, the shortest will become a new promoter.
intersect_promoters_gaps <- GenomicRanges::intersect(promoters_4500, gaps_genes) #was masked by dyplr pakage!

table(width(intersect_promoters_gaps))

# 225  465  880 2911 3630 4351 4500 
#   1    1    1    1    1    1    3 

intersect_gaps_promoters <- GenomicRanges::intersect(gaps_genes, promoters_4500)

setdif_gaps_promoters <- GenomicRanges::setdiff(gaps_genes, promoters_4500)
setdif_promoters_gaps <- GenomicRanges::setdiff(promoters_4500, gaps_genes)

granges(Small_genes[which(strand(Small_genes) == "-")])[1:6]
granges(promoters_4500[which(strand(promoters_4500) == "-")])[1:6]
granges(gaps_genes[which(strand(gaps_genes) == "-")])[1:6]
granges(intersect_promoters_gaps[which(strand(intersect_promoters_gaps) == "-")])[1:6]

par(mfrow=c(4,1))
plotRanges(Small_genes[which(strand(Small_genes) == "+")],xlim=c(40000,50000))
plotRanges(promoters_4500[which(strand(promoters_4500) == "+")],xlim=c(40000,50000))
plotRanges(gaps_genes[which(strand(gaps_genes) == "+"  & seqnames(gaps_genes) == "1")],xlim=c(40000,50000))
plotRanges(intersect_promoters_gaps[which(strand(intersect_promoters_gaps) == "+")],xlim=c(40000,50000))
#plotRanges(intersect_gaps_promoters[which(strand(intersect_gaps_promoters) == "+")],xlim=c(0,50000))


par(mfrow=c(4,1))
plotRanges(Small_genes[which(strand(Small_genes) == "-")],xlim=c(40000,50000))
plotRanges(promoters_4500[which(strand(promoters_4500) == "-")],xlim=c(40000,50000))
plotRanges(gaps_genes[which(strand(gaps_genes) == "-" & seqnames(gaps_genes) == "1")],xlim=c(40000,50000))
plotRanges(intersect_promoters_gaps[which(strand(intersect_promoters_gaps) == "-")],xlim=c(40000,50000))

plotRanges(setdif_promoters_gaps[which(strand(setdif_promoters_gaps) == "-")],xlim=c(0,50000))
plotRanges(setdif_gaps_promoters[which(strand(setdif_gaps_promoters) == "-")],xlim=c(0,50000))


```
#Figure for group meeting:
```{r}
genes_fwd <- Small_genes[which(strand(Small_genes) == "+")]
promoters_4500_fwd <- promoters_4500[which(strand(promoters_4500) == "+")]
gaps_genes_fwd <- gaps_genes[which(strand(gaps_genes) == "+"  & seqnames(gaps_genes) == "1")]
norm_promoters_fwd <- intersect_promoters_gaps[which(strand(intersect_promoters_gaps) == "+")]

genes_rev <- Small_genes[which(strand(Small_genes) == "-")]
promoters_4500_rev <- promoters_4500[which(strand(promoters_4500) == "-")]
gaps_genes_rev <- gaps_genes[which(strand(gaps_genes) == "-"  & seqnames(gaps_genes) == "1")]
norm_promoters_rev <- intersect_promoters_gaps[which(strand(intersect_promoters_gaps) == "-")]

png("Promoters_method_explanation_fwd.png")
par(mfrow=c(4,1))
plotRanges(genes_fwd,xlim=c(0,50000))
plotRanges(gaps_genes_fwd,xlim=c(0,50000))
plotRanges(promoters_4500_fwd,xlim=c(0,50000))
plotRanges(norm_promoters_fwd,xlim=c(0,50000))
dev.off()

png("Promoters_method_explanation_rev.png")
par(mfrow=c(4,1))
plotRanges(genes_rev,xlim=c(0,50000))
plotRanges(gaps_genes_rev,xlim=c(0,50000))
plotRanges(promoters_4500_rev,xlim=c(0,50000))
plotRanges(norm_promoters_rev,xlim=c(0,50000))
dev.off()

```

```{r}
#find, not overlaping promoter, it will be the first one. 
non_over <- promoters_4500[-queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps))]
promoters_4500[unique(queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]
intersect_promoters_gaps[unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]

#Extarct Metadata from promoters_4500
intersect_promoters_gaps_reord <- intersect_promoters_gaps[unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]

mcols(intersect_promoters_gaps_reord) <- 
  mcols(promoters_4500[unique(queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))])

```

```{r}

gene_promoters_all <-GenomicRanges::sort(c(intersect_promoters_gaps_reord,non_over), ignore.strand=TRUE)
mcols(gene_promoters_all)$type <- "promoter"
mcols(gene_promoters_all)$biotype <- "promoter"

gene_promoters_all
Small_genes

#Put this step at the end
gene_promoters_all_filter <- gene_promoters_all[width(gene_promoters_all) >= minSize]
  
par(mfrow=c(4,1))
plotRanges(Small_genes[which(strand(Small_genes) == "+")],xlim=c(0,60000))
plotRanges(gene_promoters_all[which(strand(gene_promoters_all) == "+")],xlim=c(0,60000))
plotRanges(gene_promoters_all_filter,xlim=c(0,60000))

genes_promoters <- sort(c(Small_genes, gene_promoters_all_filter), ignore.strand=TRUE)
genes_promoters[1:6, 15:17]
```

```{r}
granges(gene_promoters_all)[1:3]
granges(Small_genes)[1:3]

gene_promoters_all[1:3]
Small_genes[1:3]

GRanges(granges(gene_promoters_all), mcols(Small_genes))
mcols(Small_genes)
mcols(gene_promoters_all) <- mcols(Small_genes)
mcols(gene_promoters_all)$type <- "promoter"
mcols(gene_promoters_all)$biotype <- "promoter"


```


#Try the analysis for "+" and "-" separately to check for bugs
```{r}
positive <- Small_genes[strand(Small_genes) == "+"]
negative <- Small_genes[strand(Small_genes) == "-"]


# Define all promoters upstream
#up_subject <- promoters(subject, upstream = 5, downstream = 0)

#----- promoter sequence (1.5kbp, 3kbp, and 4.5kbp upstream
prom_size <- 4500 

#2. Define all_promoters
promoters_4500_pos <- GenomicRanges::promoters(positive, upstream = prom_size, downstream = 0)
promoters_4500_neg <- GenomicRanges::promoters(negative, upstream = prom_size, downstream = 0)

#find gaps bteween genes
gaps_genes_pos <- GenomicRanges::gaps(positive)
gaps_genes_neg <- GenomicRanges::gaps(negative)

table(width(gaps_genes_pos))
#3630 13449 17246 
#    1     1     1 


#find overlap between promoters and gaps between genes, the shortest will become a new promoter.
intersect_promoters_gaps_pos <- GenomicRanges::intersect(promoters_4500_pos, gaps_genes_pos) #was masked by dyplr pakage!
intersect_promoters_gaps_neg <- GenomicRanges::intersect(promoters_4500_neg, gaps_genes_neg)

table(width(intersect_promoters_gaps_pos))

# 3630 4500 
#   1    2 

granges(positive)
granges(promoters_4500_pos)
granges(gaps_genes_pos)
granges(intersect_promoters_gaps_pos)

par(mfrow=c(4,1))
plotRanges(positive,xlim=c(0,50000))
plotRanges(promoters_4500_pos,xlim=c(0,50000))
plotRanges(gaps_genes_pos,xlim=c(0,50000))
plotRanges(intersect_promoters_gaps_pos,xlim=c(0,50000))

par(mfrow=c(4,1))
plotRanges(negative,xlim=c(0,50000))
plotRanges(promoters_4500_neg,xlim=c(0,50000))
plotRanges(gaps_genes_neg,xlim=c(0,50000))
plotRanges(intersect_promoters_gaps_neg,xlim=c(0,50000))

#find, not overlaping promoter, it will be the first one. 
non_over_pos <- promoters_4500_pos[-queryHits(findOverlaps(promoters_4500_pos, intersect_promoters_gaps_pos))]
non_over_neg <- promoters_4500_neg[-queryHits(findOverlaps(promoters_4500_neg, intersect_promoters_gaps_neg))]

#adjust the start not to be negative
start(non_over_pos) #continue


gene_promoters_pos <-sort(c(intersect_promoters_gaps_pos,non_over_pos))
gene_promoters_neg <-sort(c(intersect_promoters_gaps_neg,non_over_neg))

par(mfrow=c(4,1))
plotRanges(positive,xlim=c(0,60000))
plotRanges(gene_promoters_pos,xlim=c(0,60000))
plotRanges(negative,xlim=c(0,60000))
plotRanges(gene_promoters_neg,xlim=c(0,60000))


gene_promoters <- sort(c(intersect_promoters_gaps_pos,non_over_pos,intersect_promoters_gaps_neg,non_over_neg))

# Filter all promoters less than 500
gene_promoters <- gene_promoters[width(gene_promoters) >= 500]

par(mfrow=c(4,1))
plotRanges(Small_genes,xlim=c(0,60000))
plotRanges(gene_promoters,xlim=c(0,60000))
plotRanges(gene_promoters_all,xlim=c(0,60000))
#The result is the same as for all genes together
```





# Troubleshooting full Annotation analysis
# Devide dataset by chromosomes

```{r}
# 1. Creat small data set

#1. Define genes
genes <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "protein_coding",]
#check
table(mcols(genes)$type)
table(mcols(genes)$biotype)

genes_chr1 <- genes[seqnames(genes) == 1]


# Define all promoters upstream
#up_subject <- promoters(subject, upstream = 5, downstream = 0)

#----- promoter sequence (1.5kbp, 3kbp, and 4.5kbp upstream
prom_size <- 4500 
minSize <- 500

#2. Define all_promoters
promoters_4500_chr1 <- GenomicRanges::promoters(genes_chr1, upstream = prom_size, downstream = 0)


#find gaps bteween genes
gaps_genes_chr1 <- GenomicRanges::gaps(genes_chr1)

   
#find overlap between promoters and gaps between genes, the shortest will become a new promoter.
intersect_promoters_gaps_chr1 <- GenomicRanges::intersect(promoters_4500_chr1, gaps_genes_chr1) #was masked by dyplr pakage!



par(mfrow=c(4,1))
plotRanges(genes_chr1,xlim=c(0,60000))
plotRanges(promoters_4500_chr1,xlim=c(0,60000))
plotRanges(gaps_genes_chr1,xlim=c(0,60000))
plotRanges(intersect_promoters_gaps_chr1,xlim=c(0,60000))

#find, not overlaping promoter, it will be the first one. 
non_over_chr1 <- promoters_4500_chr1[-queryHits(findOverlaps(promoters_4500_chr1, intersect_promoters_gaps_chr1))]
#adjust the start not to be negative
start(non_over) #continue

gene_promoters_all_chr1 <-GenomicRanges::sort(c(intersect_promoters_gaps_chr1,non_over_chr1), ignore.strand=TRUE)

#############
#############
genes_chr1[is.na(GenomicRanges::follow(gene_promoters_all_chr1, genes_chr1))]
mcols(gene_promoters_all_chr1) <- mcols(genes_chr1)
mcols(gene_promoters_all_chr1)$type <- "promoter"
mcols(gene_promoters_all_chr1)$biotype <- "promoter"

#Put this step at the end
gene_promoters_all_filter <- gene_promoters_all[width(gene_promoters_all) >= minSize]
  
par(mfrow=c(4,1))
plotRanges(Small_genes,xlim=c(0,60000))
plotRanges(gene_promoters_all,xlim=c(0,60000))
plotRanges(gene_promoters_all_filter,xlim=c(0,60000))

genes_promoters <- sort(c(genes, gene_promoters_all_filter), ignore.strand=TRUE)
```








#promoter sequence (1.5kbp, 3kbp, and 4.5kbp upstream
## 1. Define list of promoters upstream of the "genes" based on "promoter_size"
## 2. Check for the distance between genes, if smaler then promoter size and greater then "minsize" = 0.5kb, take intergenic region as promoter.
## 3. Should we include TE in analisys? NO
## 4. 


-filter promoters contain genes, redefine them as the size from the nearest gene

```{r}
## Create subsets of GRange object by featers from Anotation file
### mcols(Annotation)$type == "gene" include "genes" and TE
# to select genes, and not TE, select first for (Annotation)$type == "gene" to selesct genes+TE, then use feild "biotype" to select only for genes, by using (Annotation)$biotype == "protein_coding"

#----- promoter sequence (1.5kbp, 3kbp, and 4.5kbp upstream
prom_size <- 4500 
minSize <- 500

#1. Define genes
genes <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "protein_coding",]
#check
table(mcols(genes)$type)
table(mcols(genes)$biotype)
GRanges(genes)
#2. Define all_promoters
promoters_4500 <- promoters(genes, upstream = prom_size, downstream = 0)

#check
head(promoters_4500, 5)
head(genes)





#3. If widht of "promoters_4500" is 
```




##Algoritm
```{r}
# "genes" = genes coordinates - TE extracted from"Annotation"
genes <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "protein_coding",]

# "Not_genes" = setdiff between "Annotation" and "genes", will contain all posible promoters = "short" + "long"
notGenes <- setdiff(Annotation, genes)
granges(notGenes)
# "Promoters_long" = filter "Not_genes" >= "promoter_size" (5000/5kb)
# "Promoters_short" = filter "Not_genes" < "promoter_size" (5000/5kb)
# narrow "Promoters_long" to be the "width" of "promoter_size", with the same "stop" coordinate
resize(g, 5, fix="end")
# "Promoters_X_lenght" = combine "Promoters_short" and "Promoters_long"
# rename type to be "Promoters"
# "gens_and_promoters" = combine "gens" and Promoters
```




```{r}
## Create subsets of GRange object by featers from Anotation file
### mcols(Annotation)$type == "gene" include "genes" and TE
# to select genes, and not TE, select first for (Annotation)$type == "gene" to selesct genes+TE, then use feild "biotype" to select only for genes, by using (Annotation)$biotype == "protein_coding"
promoter_size <- 4500 
minSize <- 500

genes <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "protein_coding",]
table(mcols(genes)$type)
table(mcols(genes)$biotype)

promoters_4500 <- promoters(genes, upstream = 4500, downstream = 0)
granges(promoters_4500)
granges(genes)
table(width(promoters_4500))
table(mcols(promoters_4500)$biotype)
table(mcols(promoters_4500)$type)
head(promoters_4500)
head(genes)
```




```{r}
promoters_500to1000 <- flank(promoters(genes, upstream = 500, downstream = 0),500)
ranges(genes)


#------
# there are 2 ways of calculation TE, based on Annotation and based on TE_Annotation_HT
# 1. based on Annotaion
transposable_element <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "transposable_element"]
granges(transposable_element)
#3903 ranges

# 2. based on TE_Annotation_HT
TE <- TE_Annotation_HT[mcols(TE_Annotation_HT)$type == "transposable_element"]
granges(TE)
# DataFrame with 31189 rows and 8 columns

#-------------
# To finnd intergenic, "Total number of insertions" - hits in "gens" + "transposable elements" + "promoters"
#The sum of all together will give 100%
#---------



five_prime_UTR <- Annotation[which(mcols(Annotation)$type == "five_prime_UTR")]
three_prime_UTR <- Annotation[which(mcols(Annotation)$type == "three_prime_UTR")]
exons <- Annotation[which(mcols(Annotation)$type == "exon")]
introns <- GenomicRanges::setdiff(Annotation[mcols(Annotation)$type == "gene"], Annotation[which(mcols(Annotation)$type == "exon")])


```

```{r}
###################
prec <- precede(intersect_promoters_gaps, genes)
length(prec)
genes[prec]
prec1 <- precede(intersect_promoters_gaps, genes)
length(prec1)
genes[prec]
#################

promoters_4500[-unique(queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]

#Extarct Metadata from promoters_4500, subject = intersect_promoters_gaps
promoters_4500_reord <- GenomicRanges::sort(promoters_4500[unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))], ignore.strand=TRUE)

intersect_promoters_gaps_reord <- GenomicRanges::sort(intersect_promoters_gaps[unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))], ignore.strand=TRUE)

promoters_4500_reord[which(seqnames(promoters_4500_reord)=="5")]
intersect_promoters_gaps_reord

tail(promoters_4500_reord[,15:17], n=6)
tail(intersect_promoters_gaps_reord, n=6)
```


```{r}
mcols(intersect_promoters_gaps_reord) <- mcols(promoters_4500_reord)

intersect_promoters_gaps_reord[1:5, 15:17]


##################PROBLEM!
###################-------Find nearest gene down of promoters, subset gene list, use this subset to provide metadata

#------Sort and combine
gene_promoters_all <-GenomicRanges::sort(c(intersect_promoters_gaps,granges(non_over)), ignore.strand=TRUE)

gene_promoters_all
granges(genes)

# Change Metadata of promoter list to include original info, and change "type" and "biotype" columns to be "promoter"
mcols(gene_promoters_all) <- mcols(genes)
mcols(gene_promoters_all)$type <- "promoter"
mcols(gene_promoters_all)$biotype <- "promoter"

# Filter short regions in potential promoter
gene_promoters_all_filter <- gene_promoters_all[width(gene_promoters_all) >= minSize]
 
# # Sort and combine to final list for genes and promoters 
genes_promoters <- sort(c(genes, gene_promoters_all_filter), ignore.strand=TRUE)

```




```{r}


lost <- promoters_4500[-unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]
lost_genes <- genes[-unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]
seqnames(lost)

length(unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps))))
length(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))
length(queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))
length(unique(queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps))))
length(start(promoters_4500))

#query = promoters_4500, subject = intersect_promoters_gaps
query <- promoters_4500[unique(queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]
subject <- intersect_promoters_gaps[unique(subjectHits(findOverlaps(promoters_4500, intersect_promoters_gaps)))]
query[,15:17]
subject

c(query[start(query) %in% start(subject)], query[end(query) %in% end(subject)])
c(query[na.omit(match(start(query), start(subject)))],
query[na.omit(match(end(query), end(subject)))])

query_df = as(query, "data.frame")
subject_df = as(subject, "data.frame")

##############
prec <- precede(intersect_promoters_gaps, genes)
length(prec)
genes[prec]
##############
```



```{r}
gr0 <- GRanges(
    Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
    IRanges(c(1:9,7L), end=10),
    strand=Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    seqlengths=c(chr1=11, chr2=12, chr3=13)
)
gr <- c(gr0, gr0[7:3])
names(gr) <- LETTERS[seq_along(gr)]


## ---------------------------------------------------------------------
## B. match(), selfmatch(), %in%, duplicated(), unique()
## ---------------------------------------------------------------------
table <- gr[1:7]
match(gr, table)
match(gr, table, ignore.strand=TRUE)

gr %in% table

duplicated(gr)
unique(gr)

## ---------------------------------------------------------------------
## C. findMatches(), countMatches()
## ---------------------------------------------------------------------
findMatches(gr, table)
countMatches(gr, table)

findMatches(gr, table, ignore.strand=TRUE)
countMatches(gr, table, ignore.strand=TRUE)

gr_levels <- unique(gr)
countMatches(gr_levels, gr)

```


```{r}

par(mfrow=c(4,1))

#plotRanges(lost_genes[which(strand(lost_genes) == "+" & seqnames(lost) == "5")],xlim=c(26920000,26960000))
plotRanges(genes[which(strand(genes) == "+" & seqnames(genes) == "5")],xlim=c(26920000,26960000))
plotRanges(promoters_4500[which(strand(promoters_4500) == "+" & seqnames(promoters_4500) == "5")],xlim=c(26920000,26960000))
plotRanges(lost[which(strand(lost) == "+" & seqnames(lost) == "5")],xlim=c(26920000,26960000))
plotRanges(intersect_promoters_gaps_reord[which(strand(intersect_promoters_gaps_reord) == "+" & seqnames(intersect_promoters_gaps_reord) == "5")],xlim=c(26920000,26960000))

plotRanges(lost[which(strand(lost) == "-" & seqnames(lost) == "5")],xlim=c(26800000,27000000))
plotRanges(lost_genes[which(strand(lost_genes) == "-" & seqnames(lost) == "5")],xlim=c(26800000,27000000))

par(mfrow=c(4,1))
#plotRanges(lost_genes[which(strand(lost_genes) == "+" & seqnames(lost) == "5")],xlim=c(26920000,26960000))
plotRanges(genes[which(strand(genes) == "+" & seqnames(genes) == "5")],xlim=c(26930000,26945000))
plotRanges(promoters_4500[which(strand(promoters_4500) == "+" & seqnames(promoters_4500) == "5")],xlim=c(26930000,26945000))
plotRanges(lost[which(strand(lost) == "+" & seqnames(lost) == "5")],xlim=c(26930000,26945000))
plotRanges(intersect_promoters_gaps_reord[which(strand(intersect_promoters_gaps_reord) == "+" & seqnames(intersect_promoters_gaps_reord) == "5")],xlim=c(26930000,26945000))


##############
prec <- precede(intersect_promoters_gaps, genes)
length(prec)
gene_list <- genes[prec]

par(mfrow=c(4,1))
plotRanges(gene_list[which(strand(gene_list) == "+" & seqnames(gene_list) == "5")],xlim=c(26930000,26945000))
plotRanges(promoters_4500[which(strand(promoters_4500) == "+" & seqnames(promoters_4500) == "5")],xlim=c(26930000,26945000))
plotRanges(intersect_promoters_gaps[which(strand(intersect_promoters_gaps) == "+" & seqnames(intersect_promoters_gaps) == "5")],xlim=c(26930000,26945000))

pdf("Promoter.pdf")
par(mfrow=c(4,1))
plotRanges(genes[which(strand(genes) == "+" & seqnames(genes) == "5")],xlim=c(26000000,26200000))
plotRanges(promoters_4500[which(strand(promoters_4500) == "+" & seqnames(promoters_4500) == "5")],xlim=c(26000000,26200000))
plotRanges(gene_list[which(strand(gene_list) == "+" & seqnames(gene_list) == "5")],xlim=c(26000000,26200000))
plotRanges(intersect_promoters_gaps[which(strand(intersect_promoters_gaps) == "+" & seqnames(intersect_promoters_gaps) == "5")],xlim=c(26000000,26200000))
dev.off()
```

```{r}
par(mfrow=c(4,1))
#plotRanges(lost_genes[which(strand(lost_genes) == "+" & seqnames(lost) == "5")],xlim=c(26920000,26960000))
plotRanges(genes[which(strand(genes) == "-" & seqnames(genes) == "5")],xlim=c(26920000,26960000))
plotRanges(promoters_4500[which(strand(promoters_4500) == "-" & seqnames(promoters_4500) == "5")],xlim=c(26920000,26960000))
plotRanges(lost[which(strand(lost) == "-" & seqnames(lost) == "5")],xlim=c(26920000,26960000))
plotRanges(intersect_promoters_gaps_reord[which(strand(intersect_promoters_gaps_reord) == "-" & seqnames(intersect_promoters_gaps_reord) == "5")],xlim=c(26920000,26960000))

```

```{r}
table(queryHits(findOverlaps(genes,genes)) != queryHits(findOverlaps(genes,genes)))
o_genes <- genes[unique(queryHits(findOverlaps(genes,promoters_4500)))][,15:17]
o_prom <- promoters_4500[unique(subjectHits(findOverlaps(genes,promoters_4500)))][,15:17]


o_genes[which(strand(o_genes) == "+")]
o_prom[which(strand(o_prom) == "+")]

par(mfrow=c(2,1))
plotRanges(o_genes[which(strand(o_genes) == "+")][1:200],xlim=c(440000,500000))
plotRanges(o_prom[which(strand(o_prom) == "+")][1:200],xlim=c(440000,500000))

```




```{r}

```


```{r}

# Define all promoters upstream
#up_subject <- promoters(subject, upstream = 5, downstream = 0)

#----- promoter sequence (1.5kbp, 3kbp, and 4.5kbp upstream


#2. Define all_promoters
promoters_4500 <- GenomicRanges::promoters(Small_genes, upstream = prom_size, downstream = 0)


#find gaps bteween genes
gaps_genes <- GenomicRanges::gaps(Small_genes)
table(width(gaps_genes))
#225   465   880  2911  3630  4351  5927 13449 17246 17455 
#    1     1     1     1     1     1     1     1     1     1 
    
#find overlap between promoters and gaps between genes, the shortest will become a new promoter.
intersect_promoters_gaps <- GenomicRanges::intersect(promoters_4500, gaps_genes) #was masked by dyplr pakage!

table(width(intersect_promoters_gaps))

# 225  465  880 2911 3630 4351 4500 
#   1    1    1    1    1    1    3 



GRanges(Small_genes)[1:3]
GRanges(promoters_4500)[1:3]
GRanges(gaps_genes)[1:3]
GRanges(intersect_promoters_gaps)[1:3]

par(mfrow=c(4,1))
plotRanges(Small_genes,xlim=c(0,50000))
plotRanges(promoters_4500,xlim=c(0,50000))
plotRanges(gaps_genes,xlim=c(0,50000))
plotRanges(intersect_promoters_gaps,xlim=c(0,50000))

#find, not overlaping promoter, it will be the first one. 
non_over <- promoters_4500[-queryHits(findOverlaps(promoters_4500, intersect_promoters_gaps))]
#adjust the start not to be negative
start(non_over) #continue

gene_promoters_all <-GenomicRanges::sort(c(intersect_promoters_gaps,non_over), ignore.strand=TRUE)
mcols(gene_promoters_all) <- mcols(Small_genes)
mcols(gene_promoters_all)$type <- "promoter"
mcols(gene_promoters_all)$biotype <- "promoter"

#Put this step at the end
gene_promoters_all_filter <- gene_promoters_all[width(gene_promoters_all) >= minSize]
  
par(mfrow=c(4,1))
plotRanges(Small_genes,xlim=c(0,60000))
plotRanges(gene_promoters_all,xlim=c(0,60000))
plotRanges(gene_promoters_all_filter,xlim=c(0,60000))

genes_promoters <- sort(c(genes, gene_promoters_all_filter), ignore.strand=TRUE)
```


