---
title: "Distribution of TF binding sites on promoters"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Distribution of TF binding sites of different TF families on promoters

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(RColorBrewer)
library(gplots)

library(randomForest)
library(broom) # create dataframe from statistical tests outputs
library(gridExtra) # to save ggplots on 1 page

# library(plyr) # to round up to tearest 10; problem with some dplyr comands

library(rtracklayer)
library(Biostrings)
library(GenomicRanges)

source("~/Work/2019/Analysis/Utilities_R/Functions/My_functions.R")

set.seed(100)
```

# load DATA

```{r}
load("~/Work/2019/Analysis/Promoters/BRapa/Brapa_v3.0/promoters_analysis/data/data_to_load/Brapa_promoters_analysis_to_load.RData")

# Brapa3_genomes_full <- # Brapa and Ara orthologs
#     read.delim("~/Work/2019/Data/JIC/Brapa/reference/3genomes_AKBr_v3_1810.txt", header=FALSE, stringsAsFactors=FALSE)
#  
# Brapa3_promoters1500_annotation <- # gff3 file of annotated promoters
#     import.gff3("~/Work/2019/Analysis/Promoters/BRapa/Brapa_v3.0/promoters/Brapa_3.0_promoters_1500.gff3")
# 
# Brapa3_TF_results_scan <- read.csv("~/Work/2019/Analysis/TF_binding/Brapa_TF/Brapa_3/analysed_data/TF_scan_results_Brapa3_promoters1500_TF_family.csv")
# 
# # load("~/Work/2019/Analysis/Arabidopsis/Exploratory_analysis/Data/to_load/TF_binding_sites_distribution.RData")
# Ara_promoters1500_annotation <- import.gff3("~/Work/2019/Analysis/Promoters/promoters_1500.gff3")
# Ara_TF_results_scan <- read_csv("~/Work/2019/Analysis/TF_binding/analysed_data/TF_scan_results_promoters1500_TF_family_v1.csv" )
# 
# # result of clustering of variable genes by clust without optimisation
# clust_zscoresVariableGenes_triplicate_10TPM_prom_no_op <- read.csv("~/Work/2019/Analysis/Arabidopsis/Expression/Clust_phyton/updated_promoter/clust_output_expression/clust_zscoresVariableGenes_triplicate_10TPM_prom_no_op.csv", stringsAsFactors=FALSE)

```

# Pre-process data and combine Brapa3_genomes with Brapa TF predicted binding sites data
```{r}
Brapa3_genomes <- select(Brapa3_genomes_full, V1, "AGI" = V2, "Genome" = V4, "1" = V5, "2" = V6, "3" = V7) %>% 
  gather("homoeolog", "gene_id", -AGI, -Genome, -V1) %>%  filter(gene_id != "-")

length(unique(Brapa3_genomes$V1)) 
length(Brapa3_genomes$V1)
# [1] 23664 unique Arabidopsis genes
# [1] 34445 unique Brasicca genes

# homoeologs with no orthologs in Arabidopsis 
table(filter(Brapa3_genomes, AGI == "-")$homoeolog)
length(filter(Brapa3_genomes, AGI == "-")$homoeolog)

#    1    2    3 
# 2208 1794 1256 
#[1] 5258

# Number of annatated promoters with predicted TF binding sites
length(unique(Brapa3_TF_results_scan$sequence))
# [1] 43961

# Update TF resultes based on Brapa-Arabidopsis orthologs
Brapa3_TF_results_homoeolog <- inner_join(Brapa3_TF_results_scan, Brapa3_genomes, by = c("sequence" = "gene_id"))
dim(Brapa3_TF_results_homoeolog)
#[1] 4936837      14

length(unique(Brapa3_TF_results_homoeolog$sequence))
# [1] 33550 genes remained after merging

length(unique(Brapa3_genomes$gene_id))
length(unique(Brapa3_TF_results_scan$sequence))
length(unique(Brapa3_TF_results_homoeolog$sequence))

# [1] 34444
# [1] 43961
# [1] 33550
```

## Add distanse from TF binding site to TSS
```{r}
Brapa3_TF_results_homoeolog <- data.frame("promoter_width" = width(Brapa3_promoters1500_annotation),
          "gene_id" = mcols(Brapa3_promoters1500_annotation)$ID, 
          "strand" = strand(Brapa3_promoters1500_annotation)) %>% 
          inner_join(., select(Brapa3_TF_results_homoeolog, -strand), by = c("gene_id" = "sequence"))


# Check the number of the unique genes in the resulting table: 
length(unique(Brapa3_TF_results_homoeolog$gene_id))
# [1] 33550

# Calculate distance from TF_binding to TSS (promoter_width - TF_start)
Brapa3_TF_results_homoeolog$distance_to_TSS <- Brapa3_TF_results_homoeolog$promoter_width -
                                               Brapa3_TF_results_homoeolog$start + 1
```

# test: plot distance_to_TSS vs score or score.pct for "ARF_tnt" 
```{r}
test_2 <- Brapa3_TF_results_homoeolog %>% 
  filter(TF_family == "ARF_tnt") %>% 
  select(score.pct, distance_to_TSS) %>% 
  mutate(distance_to_TSS = round(distance_to_TSS, -1), score.pct = round(score.pct)) %>% 
  group_by(distance_to_TSS) %>% 
  mutate(N_total = n())  %>% 
  group_by(distance_to_TSS, score.pct) %>% 
  mutate(N_score = n(), N_rel = (N_score/N_total)*100) %>% ungroup() %>% 
  select(-N_total, -N_score) %>% 
  unique() 


###################

test_5 <- Brapa3_TF_results_homoeolog %>% 
  filter(TF_family == "ARF_tnt") %>% 
  select(score, distance_to_TSS) %>% 
  mutate(distance_to_TSS = round(distance_to_TSS, -1), score = round(score)) %>% 
  group_by(distance_to_TSS) %>% 
  mutate(N_total = n())  %>% 
  group_by(distance_to_TSS, score) %>% 
  mutate(N_score = n(), N_rel = (N_score/N_total)*100) %>% ungroup() %>% 
  select(-N_total, -N_score) %>% 
  unique() 


test_2 %>% 
  ggplot() +
  geom_point(aes(x = distance_to_TSS, y = score.pct, alpha = N_rel), colour = "blue")

test_5 %>% 
  ggplot() +
  geom_point(aes(x = distance_to_TSS, y = score, alpha = N_rel), colour = "blue")

rm(test_2, test_5)
# When using score.pct, the enrichment at the end of the promoter is lost. Why?
```

# Plot Brapa_distance_to_TSS_vs_TF_binding_score
```{r}
TF_family <- unique(Brapa3_TF_results_homoeolog$TF_family)


 # Check max score for family:   
 Brapa3_TF_results_homoeolog %>% 
  group_by(TF_family) %>% 
  summarise(TF_max_score = max(max.score)) %>% 
   arrange(desc(TF_max_score))
 
 # Plot results for all families:
 ## fixed y
 plot <- lapply(TF_family, function(x)
   Brapa3_TF_results_homoeolog %>% 
  filter(TF_family == x) %>% 
  ggplot(aes(x = distance_to_TSS, y = score, alpha = 0.1, colour = "blue")) +
  geom_point(colour = "blue", alpha = 0.25) +
  labs(title = x) + 
  theme_minimal() +
  theme(legend.position = "none") +
  scale_y_continuous(limits = c(0, 70))
   ) 
 
 ## free y
 plot1 <- lapply(TF_family, function(x)
   Brapa3_TF_results_homoeolog %>% 
  filter(TF_family == x) %>% 
  ggplot(aes(x = distance_to_TSS, y = score, alpha = 0.1, colour = "blue")) +
  geom_point(colour = "blue", alpha = 0.25) +
  labs(title = x) + 
  theme_minimal() +
  theme(legend.position = "none")
   ) 
 
 plot[[1]]
```

# DONT RUN!
```{r eval=FALSE, include=FALSE}
j <- 1
k <- 16
 for (i in 1:4){
  ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_fixed_y_", i, ".jpeg"), marrangeGrob(grobs = plot[j:k], nrow=4, ncol=4))
 j <- j+16
 k <- k + 16
 }
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_fixed_y_", 1, ".jpg"), marrangeGrob(grobs = plot[1:16], nrow=4, ncol=4))
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_fixed_y_", 2, ".jpg"), marrangeGrob(grobs = plot[17:32], nrow=4, ncol=4))
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_fixed_y_", 3, ".jpg"), marrangeGrob(grobs = plot[33:48], nrow=4, ncol=4))
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_fixed_y_", 4, ".jpg"), marrangeGrob(grobs = plot[49:56], nrow=4, ncol=4))


 
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_free_y_", 1, ".jpg"), marrangeGrob(grobs = plot1[1:16], nrow=4, ncol=4))
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_free_y_", 2, ".jpg"), marrangeGrob(grobs = plot1[17:32], nrow=4, ncol=4))
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_free_y_", 3, ".jpg"), marrangeGrob(grobs = plot1[33:48], nrow=4, ncol=4))
ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_free_y_", 4, ".jpg"), marrangeGrob(grobs = plot1[49:56], nrow=4, ncol=4))
```


####--------##########

# Plot: Brapa_distance_to_TSS_vs_TF_binding_score_by_groups (Ancestral Chromosomal blocks)
```{r}
plot2 <- lapply(TF_family, function(x)
   Brapa3_TF_results_homoeolog %>%
   filter(TF_family == x) %>%
   ggplot(aes(x = distance_to_TSS, y = score)) +
   geom_point(colour = "blue", alpha = 0.25) +
   labs(title = x) +
   theme_minimal() +
   theme(legend.position = "none") +
   facet_wrap( ~ as.factor(Genome), ncol = 5) 
   ) 
 

j <- 1
k <- 4
 for (i in 1:(length(TF_family)/4)){
  ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_by_groups/Brapa_distance_to_TSS_vs_TF_binding_score_by_groups", i, ".jpg"),
         marrangeGrob(grobs = plot2[j:k], nrow=2, ncol=2))
 j <- j + 4
 k <- k + 4
 }
```

######-----ADD-------#######
####-----test------####
# Summarise DATA: Brapa_distance_to_TSS_vs_TF_binding_score_by_groups (Ancestral Chromosomal blocks)
# Create bar_plot

####----------------########
# Plot: Brapa_distance_to_TSS_vs_TF_binding_score_by_homoeolog 
```{r}
# test:
Brapa3_TF_results_homoeolog %>%
   filter(TF_family == "BBRBPC_tnt") %>%
   ggplot(aes(x = distance_to_TSS, y = score)) +
   geom_point(colour = "blue", alpha = 0.25) +
   labs(title = TF_family[1]) +
   theme_minimal() +
   theme(legend.position = "none") +
   facet_wrap( ~ as.factor(homoeolog), ncol = 3) + 
      theme(strip.text.x = element_text(size = 8), panel.spacing = unit(0.1, "lines")) 
 

#  theme(strip.text = element_text(face="bold", size=8),
        
 plot_3 <- lapply(TF_family, function(x)
   Brapa3_TF_results_homoeolog %>%
   filter(TF_family == x) %>%
   ggplot(aes(x = distance_to_TSS, y = score)) +
   geom_point(colour = "blue", alpha = 0.25) +
   labs(title = x) +
   theme_minimal() +
   theme(legend.position = "none")  +
   facet_wrap( ~ as.factor(homoeolog), ncol = 3)  + 
      theme(strip.text.x = element_text(size = 8), panel.spacing = unit(0.1, "lines"))  
   ) 
 
# dir.create("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_by_homoeolog")
j <- 1
k <- 8
 for (i in 1 : (length(TF_family)/8)){
  ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_by_homoeolog/Brapa_distance_to_TSS_vs_TF_binding_score_by_homoeolog", i, ".jpg"),
         marrangeGrob(grobs = plot_3[j:k], nrow=4, ncol=2))
 j <- j + 8
 k <- k + 8
 }

####---log(y)
 plot_4 <- lapply(TF_family, function(x)
   Brapa3_TF_results_homoeolog %>%
   filter(TF_family == x) %>%
   ggplot(aes(x = distance_to_TSS, y = log(score))) +
   geom_point(colour = "blue", alpha = 0.25) +
   labs(title = x) +
   theme_minimal() +
   theme(legend.position = "none")  +
   facet_wrap( ~ as.factor(homoeolog), ncol = 3)  + 
      theme(strip.text.x = element_text(size = 8), panel.spacing = unit(0.1, "lines"),
            title = element_text(size = 10))
   ) 
 
# dir.create("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_by_homoeolog")
j <- 1
k <- 8
 for (i in 1 : (length(TF_family)/8)){
  ggsave(paste0("Plots/Brapa_distance_to_TSS_vs_TF_binding_score_by_homoeolog/Brapa_distance_to_TSS_vs_TF_binding_log_score_by_homoeolog", i, ".jpg"),
         marrangeGrob(grobs = plot_4[j:k], nrow=4, ncol=2))
 j <- j + 8
 k <- k + 8
 }

plot_4[[2]]
```