---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(rtracklayer)
library(GenomicRanges)
library(tidyverse)
```

```{r}
Annotation_file <- "Brapa_1.0/Brassica_rapa.Brapa_1.0.42.chr.gff3"
prom_size <- 1500 #Desired promoter size
minSize <- 500 #Minimal promoter size

output_promoter_file <- paste0("promoters_", prom_size, ".gff3")
output_genes_promoters_file <- paste0("genes_promoterss_", prom_size, ".gff3")

Annotation <- import.gff3(Annotation_file)
#GRanges object with 483806 ranges and 19 metadata columns
seqnames(Annotation)



#Add chromasome sizes
chrom_size_Brapa_1.0 <- width(Annotation[strand(Annotation) == "*"])
seqlengths(Annotation) <- as.vector(chrom_size_Brapa_1.0) #Essential info for following calculations
seqlengths(Annotation)

#      A01      A02      A03      A04      A05      A06      A07      A08      A09      A10 
# 28605136 27846329 31715688 18965343 23939834 26271742 22586724 21594650 37120481 17594535 

# Define genes, No TE
genes <- Annotation[mcols(Annotation)$type == "gene" & mcols(Annotation)$biotype == "protein_coding",]
#GRanges object with 39607 ranges and 19 metadata columns
```


#Find coordinates of promoters with updated width, exclude genes inside the poromoter and update promoter "start"
```{r}
# Define all potential promoters and trim out of border regions
promoters <- trim(GenomicRanges::promoters(genes, upstream = prom_size, downstream = 0))

# filter out all potential promoters which include genes, and change size of promoters if genes are too close, filter for short promoters (lee than minimal length):
# 1------find gaps bteween genes
gaps_genes <- GenomicRanges::gaps(genes)

#------find overlap between promoters and gaps between genes, the shortest will become a new promoter.
intersect_promoters_gaps <- GenomicRanges::intersect(promoters, gaps_genes)

# Filter short regions (less then minSize) in potential promoter
intersect_promoters_gaps_filter <- intersect_promoters_gaps[width(intersect_promoters_gaps) >= minSize]

#------FInd genes following promoters
#"precede(query, subject)" returns the index of the interval in subject that is directly preceded by the "query" range. Overlaping regions are excluded
#gene_list <- genes[precede(intersect_promoters_gaps, genes)]
gene_list <- genes[precede(intersect_promoters_gaps_filter, genes)]
length(gene_list) == length(intersect_promoters_gaps_filter) #should be TRUE

#------Update Metadata of intersect_promoters_gaps, and change "ID", "type" and "biotype" columns to be "promoter". "ID" should be unique identifier!!
mcols(intersect_promoters_gaps_filter) <- mcols(gene_list) 

mcols(intersect_promoters_gaps_filter)$type <- "promoter"
mcols(intersect_promoters_gaps_filter)$biotype <- "promoter"
mcols(intersect_promoters_gaps_filter)$ID <- str_replace(mcols(intersect_promoters_gaps_filter)$ID, "gene", "promoter")

 #------Sort and combine to final list for genes and promoters 
genes_promoters <- sort(c(gene_list, intersect_promoters_gaps_filter), ignore.strand=TRUE)


#Save annatation files for promoters and gene_promoters
rtracklayer::export.gff3(intersect_promoters_gaps_filter, output_promoter_file, format = "GFF3") #GRanges object with 38688 ranges and 19 metadata columns
rtracklayer::export.gff3(genes_promoters, output_genes_promoters_file, format = "GFF3")

#End
```

# To extract the sequence of promoters continue with bedtools using following comands:
samtools faidx Brapa_1.0/Brassica_rapa.Brapa_1.0.dna.toplevel.fa

bedtools getfasta -fi Brapa_1.0/Brassica_rapa.Brapa_1.0.dna.toplevel.fa -bed promoters_1500.gff3 -s -tab -fo Brapa_promoters1500_tab.fa
bedtools getfasta -fi Brapa_1.0/Brassica_rapa.Brapa_1.0.dna.toplevel.fa -bed promoters_1500.gff3 -s -fo Brapa_promoters1500.fa
# -s to take the directoinality into acount
# -fo to define the output file name
# -tab for tab delimanated output

*The "start" coordinates of output file should be updated by +1, since bed file is 0-based start position (gff is 1-based). There is no need to update gff file before using bedtools, it knows how to deal with gff.


#######----------#############
#Investigation of promoters:
##Prepare the list of promoters (with updated coordinates) withot filtering by min_size
```{r}
length(genes)
# [1] 39607
length(intersect_promoters_gaps)
# [1] 39488
length(intersect_promoters_gaps_filter)
# [1] 38688
length(genes)-length(intersect_promoters_gaps_filter)
# [1] 919

```

######----The END-----#######


*Haven't been done!!!!!!!*

```{r}
# intersect_promoters_gaps - all updated promoters before filtering by minsize

#------FInd genes following promoters
#"precede(query, subject)" returns the index of the interval in subject that is directly preceded by the "query" range. Overlaping regions are excluded
#gene_list <- genes[precede(intersect_promoters_gaps, genes)]
gene_list_non_filter <- genes[precede(intersect_promoters_gaps, genes)]
length(gene_list_non_filter) == length(intersect_promoters_gaps) #should be TRUE

#------Update Metadata of intersect_promoters_gaps, and change "ID", "type" and "biotype" columns to be "promoter". "ID" should be unique identifier!!
mcols(intersect_promoters_gaps) <- mcols(gene_list_non_filter) 
mcols(intersect_promoters_gaps)$ID <- str_replace(intersect_promoters_gaps$ID, "gene:", "")
mcols(gene_list_non_filter)$ID <- str_replace(gene_list_non_filter$ID, "gene:", "")
mcols(intersect_promoters_gaps)$type <- "promoter"
mcols(intersect_promoters_gaps)$biotype <- "promoter"

#Save annatation files for promoters without filtering
rtracklayer::export.gff3(intersect_promoters_gaps, paste0("non_filtered_", output_promoter_file), format = "GFF3")
rtracklayer::export.gff3(gene_list_non_filter, paste0("non_filtered_genes_", output_promoter_file), format = "GFF3")


```

Find the promoters of the genes with promoter size less then 500
```{r}
intersect_promoters_gaps_filtered_out <- intersect_promoters_gaps[width(intersect_promoters_gaps) < minSize]

#------FInd genes following promoters
#"precede(query, subject)" returns the index of the interval in subject that is directly preceded by the "query" range. Overlaping regions are excluded
gene_list_filtered_out <- genes[precede(intersect_promoters_gaps_filtered_out, genes)]
length(gene_list_filtered_out) == length(intersect_promoters_gaps_filtered_out) #should be TRUE

#------Update Metadata of intersect_promoters_gaps, and change "ID", "type" and "biotype" columns to be "promoter". "ID" should be unique identifier!!
mcols(intersect_promoters_gaps_filtered_out) <- mcols(gene_list_filtered_out) 
mcols(intersect_promoters_gaps_filtered_out)$ID <- str_replace(intersect_promoters_gaps_filtered_out$ID, "gene:", "")
mcols(gene_list_filtered_out)$ID <- str_replace(gene_list_filtered_out$ID, "gene:", "")
mcols(intersect_promoters_gaps_filtered_out)$type <- "promoter"
mcols(intersect_promoters_gaps_filtered_out)$biotype <- "promoter"

#Save annatation files for promoters without filtering
rtracklayer::export.gff3(intersect_promoters_gaps_filtered_out, paste0("500_filtered_out_", output_promoter_file), format = "GFF3")
rtracklayer::export.gff3(gene_list_filtered_out, paste0("500_filtered_out_genes_", output_promoter_file), format = "GFF3")

```

