---
title: "Untitled"
author: "Brestovitsky A"
date: "7 February 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##IRanges

```{r}
library(IRanges)
starts <- floor(runif(n=50,min=50,max=150))
ends <- starts + floor(runif(n=50,min=1,max=25))
query <- IRanges(start = starts, end=ends)
query <- sort(query)
seq_len(length(query))

source("~/Work/2019/Analysis/Promoter/scripts/plotRanges.R")
#useful function to plot Ranges
plotRanges


#to save as data frame
query_df = as(query, "data.frame")
subject_df = as(subject, "data.frame")

```

Intra-range operations:
```{r}

shift(query,5)
plotRanges(query)
plotRanges(shift(query,5))
plotRanges(resize(query,4))
```


Inter-range-operations:
```{r}
plotRanges(reduce(query))
plot(coverage(query))
```


Comparisons of IRanges objects:"Query" and "Subject"
```{r}
start2 <- seq(30,160,by=10)
subject <- IRanges(start = start2, width=sample(1:10,length(start2),replace=T))
ov <- findOverlaps(query, subject, type="within", maxgap=1L)
IRanges::overlapsRanges(query, subject, ov)
par(mfrow=c(3,1))
plotRanges(query,xlim=c(0,150))
plotRanges(subject,xlim=c(0,150))
plotRanges(gaps(subject),xlim=c(0,150))
```



```{r}
x <- GRanges("chr1", IRanges(c(2, 9) , c(7, 19)), strand=c("+", "-"))
y <- GRanges("chr1", IRanges(5, 10), strand="-") 

union(x, y)
union(x, y, ignore.strand = TRUE)

x
y
intersect(x, y)
intersect(x, y, ignore.strand=TRUE)
x
y
setdiff(x, y)
setdiff(y, x)
setdiff(x, y, ignore.strand=TRUE)

```


"precede(query, subject)" returns the index of the interval in subject that is directly preceded by the query range. Overlaping regions are excluded
```{r}
starts <- floor(runif(n=50,min=50,max=150))
ends <- starts + floor(runif(n=50,min=1,max=25))
query1 <- IRanges(start = starts, end=ends)
query1 <- sort(query1)

start2 <- seq(30,160,by=10)
subject1 <- IRanges(start = start2, width=sample(1:10,length(start2),replace=T))
prec <- precede(query1, subject1)
foll <- follow(query1, subject1)
query[1:5]
subject1[1:5]
subject1[prec[1:5]] #subject is view in IRanges
subject1[foll[1:5]]
nearest(query,subject1)
par(mfrow = c(4,1))
plotRanges(query,xlim=c(0,150))
plotRanges(subject1,xlim=c(0,150))
plotRanges(which(subject1[prec[]],xlim=c(0,150)))
plotRanges(subject1[foll],xlim=c(0,150))

prec
foll
length(subject)
length(query)
```

The degree of overlap between query and subject:
The "findOverlaps" -The returned object stored the indices of the query and subject ranges that overlap. 
The countOverlaps is counting the number of ranges in the query overlap with the subject. 
The actual ranges where an overlap occurs (or does not occur), we can use the "intersect"
and "setdiff" functions.

```{r}
olaps <- findOverlaps(query, subject)
queryHits(olaps)
countOverlaps(query, subject)
countOverlaps(subject, query)
subsetByOverlaps(query, subject)
setdiff(query,subject)


ov <- findOverlaps(query, subject, type="within", maxgap=1L)
IRanges::overlapsRanges(query, subject, ov)

par(mfrow=c(4,1))
plotRanges(intersect(query,subject),xlim=c(0,200))
plotRanges(setdiff(query,subject),xlim=c(0,200))
plotRanges(subsetByOverlaps(query, subject),xlim=c(0,200))
plotRanges(IRanges::overlapsRanges(query, subject, ov),xlim=c(0,200))
```

##GRanges

```{r}
library(GenomicRanges)
gr <- GRanges(seqnames=sample(LETTERS[1:3],length(query),replace=T), ranges=query)
gr <- sort(gr)
gr2 <- GRanges(seqnames=sample(LETTERS[1:3],length(subject),replace=T), ranges=subject)
gr2 <- sort(gr2)
shift(gr,5)
reduce(gr)
gr[1:4]
gaps(gr)
gaps(gr2)
coverage(gr)
coverage(gr2)

par(mfrow=c(3,1))
plotRanges(gaps(gr),xlim=c(0,200))
plotRanges(gr,xlim=c(0,200))
```
genomic coordinates | metadata
```{r}
gr <- GRanges(
    seqnames = Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
    ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
    strand = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    gene = Rle(c("-", "+", "*", "+", "-"), c(1, 2, 2, 3, 2)),
    score = 1:10,
    GC = seq(1, 0, length=10))
gr
```
working with genomic ranges/coordinates
```{r}
seqnames(gr)
ranges(gr)
strand(gr)
granges(gr) #without metadata
start(gr)
end(gr)
names(gr)
```

working with metadat:
```{r}
mcols(gr)
mcols(gr)$score
#to add info about chromosome length 
seqlengths(gr) <- c(249250621, 243199373, 198022430)
seqlengths(gr)
```
# Splitting and combining GRanges objects
```{r}
sp <- split(gr, rep(1:2, each=5))
sp
c(sp[[1]], sp[[2]], gr)
```

#Subsetting GRanges objects
```{r}
gr[2:3] #by rows index
gr[2:3, "GC"] #by raw index and name of metadata column
gr[2:3, 2]

#the second row of a GRanges object is replaced with the first row of gr
singles <- split(gr, names(gr))
grMod <- gr
grMod[2] <- singles[[1]]
head(grMod, n=3)
#or
grMod[2] <- grMod[3]
head(grMod, n=3)
```

#Methods to repeat, reverse, or select specific portions of GRanges objects.

```{r}
rep(singles[[2]], times = 3)
rep(gr[2], times = 3)
rev(gr)
head(gr, n=3)
tail(gr, n=2)
window(gr, start = 3, end = 5)
gr[IRanges(start=c(2,7), end=c(3,9))]
```

#Basic interval operations for GRanges objects

```{r}
g <- gr[1:3]
g <- append(g, singles[[10]]) #vector merging
start(g)
end(g)
width(g)
range(g)
ranges(g)
```

#intra-range methods
operate on each element of a GRanges object independent of the other 
# ?"intra-range-methods" - summarizes these methods
```{r}
g
flank(g, 10)
flank(g, 10, start = FALSE)
shift(g, 5)
g
resize(g, 5)
resize(g, 5, fix="start")
resize(g, 5, fix="end")
g
```

#inter-range methods 
involve comparisons between ranges in a single GRanges object
```{r}
reduce(g) #align the ranges and merge overlapping ranges
gaps(g) #gaps between the ranges + reduced version of the ranges
g 
disjoin(g) #a collection of non-overlapping ranges-don't understand it!
coverage(g) #quantifies the degree of overlap for all the ranges
g
par
par(mfrow=c(2,1))
plotRanges(g, xlim=c(0,200))
plotRanges(gaps(g),xlim=c(0,200))
plotRanges

```

#between-range methods - Interval set operations for GRanges objects
```{r}
g2 <- head(gr, n=2)
g2
g
union(g, g2) #combine together without repetition
intersect(g, g2) #common ranges
setdiff(g, g2) #different ranges
overlapsRanges(query, subject, ov)
```



### Try to simulate promoters problem

```{r}
start2 <- seq(30,160,by=10)
subject <- IRanges(start = start2, width=sample(1:10,length(start2),replace=T))

up_subject <- promoters(subject[1:5], upstream = 5, downstream = 0)[1:5]
gaps_subject <- gaps(subject[1:5])
union_u_g <- union(up_subject, gaps_subject)
intersect_u_g <- GenomicRanges::intersect(up_subject, gaps_subject) #was masked by dyplr pakage!
diff_u_g <- GenomicRanges::setdiff(up_subject, gaps_subject)
diff_g_u <- GenomicRanges::setdiff(up_subject, gaps_subject)

start(gaps_subject[which(width(gaps_subject) > 3)]) <- end(gaps_subject[which(width(gaps_subject) > 3)])

subject[1:3]
up_subject[1:3]
gaps_subject [1:3]

par(mfrow=c(4,1))
plotRanges(subject[1:5],xlim=c(0,100))
plotRanges(up_subject,xlim=c(0,100))
plotRanges(gaps_subject,xlim=c(0,100))
plotRanges(intersect_u_g,xlim=c(0,100))


par(mfrow=c(4,1))
plotRanges(subject[1:5],xlim=c(0,100))
plotRanges(up_subject,xlim=c(0,100))
plotRanges(gaps_subject,xlim=c(0,100))
plotRanges(union_u_g,xlim=c(0,100))


par(mfrow=c(4,1))
plotRanges(up_subject,xlim=c(0,100))
plotRanges(gaps_subject,xlim=c(0,100))
plotRanges(diff_u_g,xlim=c(0,100))
plotRanges(diff_g_u,xlim=c(0,100))
```



```{r}
start2 <- seq(30,160,by=10)
subject <- IRanges(start = start2, width=sample(1:10,length(start2),replace=T))

# Define all promoters upstream
up_subject <- promoters(subject, upstream = 5, downstream = 0)
#find gaps bteween genes
gaps_subject <- gaps(subject)
#find overlap between promoters and gaps between genes, the shortest will become a new promoter.
intersect_u_g <- GenomicRanges::intersect(up_subject, gaps_subject) #was masked by dyplr pakage!

table(width(intersect_u_g))
# filter intersect_u_g by width <- short promoters
#intersect_u_g_f <- intersect_u_g[width(intersect_u_g) < 5]

# subject[1:3]
# up_subject[1:3]
# gaps_subject [1:3]
# intersect_u_g[1:3]

# par(mfrow=c(4,1))
# plotRanges(subject,xlim=c(0,200))
# plotRanges(up_subject,xlim=c(0,200))
# plotRanges(gaps_subject,xlim=c(0,200))
# plotRanges(intersect_u_g,xlim=c(0,200))

#find, not overlaping promoter, it will be the first one. 
non_over <- up_subject[-queryHits(findOverlaps(up_subject, intersect_u_g))]
#adjust the start not to be negative
start(non_over) #continue


subject_promoter <- sort(c(intersect_u_g,non_over))

par(mfrow=c(4,1))
plotRanges(subject,xlim=c(0,200))
plotRanges(subject_promoter,xlim=c(0,200))
```






